### ISSUE TITLE
IndexOutOfBoundsException in RecordParser with HttpclientResponse

### ISSUE THREAD

**Issue by poryee** [eclipse-vertx/vert.x#4811]
### when streaming vertx httpclientresponse to record parser im faced with following kind of error

 ERROR [io.ver.cor.imp.ContextBase] (vert.x-eventloop-thread-1) Unhandled exception [Error Occurred After Shutdown]:
java.lang. Index0utOfBoundsException: writerIndex(40) + minWritableBytes(8192) exceeds maxCapacity(40): UnpooledslicedByteBuf(ridx: 0, widx: 48,
cap: 40/48, unwrapped: UnpooledByteBufAllocator$InstrumentedunpooledUnsafeHeapByteBuf(ridx: 42, widx: 42, cap: 64))
at io.netty.buffer.AbstractByteBuf.ensure₩ritableø(AbstractByteBuf. java:294)
at io.netty.buffer.AbstractByteBuf.ensureWritable (AbstractByteBuf. java:280)
at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf. java:1103)
at io.vertx.core.buffer.impl.BufferImpl.appendBuffer (BufferImpl. java:268)
I
at io.vertx.core.parsetools.impl.RecordParserImpl. handle (RecordParserImpl. java:283)
at io.vertx.core.parsetools.impl.RecordarserImpl. handle(RecordParserImpl. java:27)
at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal. java:264)
at io.vertx.core.http.impl.HttpEventHandler.handleChunk(HttpEventHandler. java:51)
at io.vertx.core.http.impl.HttpclientResponseImpl. handleChunk(HttpClientResponseImpl. java:239)
at io.vertx.core.http.impl.HttplxClientConnection$streamImpl.lambdaŞnew$0(HttplxClientConnection. iava:452)
at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer. java:255)
at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
at io.vertx.core.http.impl.HttplxClientConnection$streamImpl.handleChunk(HttplxClientConnection. java:704)
at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext. java:76)
at io.vertx.core.impl.ContextBase.execute(ContextBase. java:232)

I believe a similar ticket was resolved in version 3.9.3
#3546 

### Version
I tried both 4.3.7 and 4.4.4

It only happens when i stream large amount of data
Perfectly fine on small response.

**Comment by tsegismont** [comment]
Thank for the report @poryee . Can you provide a small reproducer?

**Comment by poryee** [comment]
Hello unfortunately my company policy doesnt allow me to copy code out. But i have resolved the problem by providing my own buffer into the parser.handle method.

**Comment by paloliska** [comment]
Hello @poryee I am using quarkus 3.5.0, which is using vertx 4.4.5, and I got the same exception: `
java.lang.IndexOutOfBoundsException: writerIndex(72) + minWritableBytes(137) exceeds maxCapacity(72): UnpooledSlicedByteBuf(ridx: 0, widx: 72, cap: 72/72, unwrapped: UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf(ridx: 78, widx: 78, cap: 100))
	at io.netty.buffer.AbstractByteBuf.ensureWritable0(AbstractByteBuf.java:294)
	at io.netty.buffer.AbstractByteBuf.ensureWritable(AbstractByteBuf.java:280)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1103)
	at io.vertx.core.buffer.impl.BufferImpl.appendBuffer(BufferImpl.java:268)
	at io.vertx.core.parsetools.impl.RecordParserImpl.handle(RecordParserImpl.java:283)
	at io.vertx.core.parsetools.impl.RecordParserImpl.handle(RecordParserImpl.java:27)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
	at io.vertx.core.http.impl.HttpEventHandler.handleChunk(HttpEventHandler.java:51)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleChunk(HttpClientResponseImpl.java:239)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:452)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:255)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleChunk(Http1xClientConnection.java:704)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:76)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:174)
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseChunk(Http1xClientConnection.java:918)
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:811)
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:778)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1471)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1334)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1383)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:509)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
`

How was this issue fixed? Is this behaviour regresion or issue is not fixed in 4.4.5? Is there something that could be done to mitigate this prior fixing?

**Comment by paloliska** [comment]
@tsegismont is this not-a-bug? Is it some misconfiguration? Or is there something that could be done on vertx / quarkus / my side to prevent this?

**Comment by tsegismont** [comment]
@paloliska I've changed the status because we compute release notes from GH issues and pull requests info.

If you can share a small reproducer, even with quarkus, I'll take a look

**Comment by poryee** [comment]
Hello, if u still havr problem can try this. Note u need to format it urself im typing thru my phone.

Request.send()
.onsuccess(resp->{
Recordparser parser = recordparser.newdelimited("\r\n");
// i have a multi on top that emits buffer
Parser.handler(em::emit)
//this is the fix
Resp.handler(buffer->{
Parser.handle(buffer.copy));
});
//ur exception handler
});

So technically it was changes due to how netty does abstractbytebuf. We pass in the copy of the buffer which recordparser can append to or something if i remember correctly. Hope this helps

**Comment by paloliska** [comment]
@tsegismont I am not able to reproduce the problem on reproducer. See my question in quarkus repo: [36940](https://github.com/quarkusio/quarkus/discussions/36940)

**Comment by tsegismont** [comment]
@paloliska have you tried the workaround proposed by @poryee ?

**Comment by paloliska** [comment]
> @paloliska have you tried the workaround proposed by @poryee ?

I dont, I am using quarkus, so to use it I have to modify quarkus code...

**Comment by paloliska** [comment]
How you @tsegismont, @poryee any ideas how to reproduce this on quarkus? Or even on vertx, can you describe the scenario?

**Comment by paloliska** [comment]
> How is the `RecordParser` paired to the `HttpRequest` then, is it done by a Quarkus extension?
> 
> It would help if you could share some snippets
It is done in resteasy here: 
[MultiInvoker:177](https://github.com/quarkusio/quarkus/blob/8948da3d3c8259efb4cb9d9be1ce5124dc99687e/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/MultiInvoker.java#L177
)

**Comment by paloliska** [comment]
in my code I am calling 
`
val proc = UnicastProcessor.create()
...
proc.onNext(eventMessage)
`
and `proc` id returned in rest endpoint

**Comment by tsegismont** [comment]
@paloliska I've sent #4960 which avoids growing the initial buffer until the delimiter (or the fixed size) is reached.

It is a better not to alter the input anyway. Nevertheless, I would like to understand how a buffer with such  low capacity can be created. If you can't provide a reproducer, can you describe what exactly your code is doing? It seems a client (which one?) is used to query an SSE endpoint, but I'm not sure.

**Comment by tsegismont** [comment]
@paloliska I've been able to create a reproducer: the client and the server should communicate over an encrypted connection.

In this case, `VertxHandler#safeHandler` does not copy the received buffer because it's not pooled.

https://github.com/eclipse-vertx/vert.x/blob/becb722bb27977a24d1a4ca988094a4ec510b168/src/main/java/io/vertx/core/net/impl/VertxHandler.java#L57-L72

These buffers are created at the encryption support layer and have a limited, low capacity.

I've added tests to #4960 that fail with the same exception without the fix in `RecordParser` (copy initial buffer).
